src/ca/idexchange.rs

```diff
@@ -0,0 +1,1197 @@
+//! Out of band exchange messages.
+//!
+//! Support for the RFC8183 out-of-band setup requests and responses
+//! used to exchange identity and configuration between CAs and their
+//! parent CA and/or RPKI Publication Servers.
+
+use std::borrow;
+use std::convert::TryFrom;
+use std::fmt;
+use std::io;
+use std::path::PathBuf;
+use std::str::from_utf8;
+use std::str::FromStr;
+use std::sync::Arc;
+
+use log::debug;
+use serde::{Deserialize, Deserializer, Serialize, Serializer};
+
+use crate::ca::idcert::IdCert;
+use crate::repository::x509::{Time, ValidationError};
+use crate::uri;
+use crate::xml;
+use crate::xml::decode::{Error as XmlError, Name};
+
+// Constants for the RFC 8183 XML
+const VERSION: &str = "1";
+const NS: &[u8] = b"http://www.hactrn.net/uris/rpki/rpki-setup/";
+
+const CHILD_REQUEST: Name = Name::qualified(NS, b"child_request");
+const CHILD_BPKI_TA: Name = Name::qualified(NS, b"child_bpki_ta");
+
+const PUBLISHER_REQUEST: Name = Name::qualified(NS, b"publisher_request");
+const PUBLISHER_BPKI_TA: Name = Name::qualified(NS, b"publisher_bpki_ta");
+
+const PARENT_RESPONSE: Name = Name::qualified(NS, b"parent_response");
+const PARENT_BPKI_TA: Name = Name::qualified(NS, b"parent_bpki_ta");
+const PARENT_REFERRAL: Name = Name::qualified(NS, b"referral");
+const PARENT_OFFER: Name = Name::qualified(NS, b"offer");
+
+const REPOSITORY_RESPONSE: Name = Name::qualified(NS, b"repository_response");
+const REPOSITORY_BPKI_TA: Name = Name::qualified(NS, b"repository_bpki_ta");
+
+//------------ Handle --------------------------------------------------------
+
+// Some type aliases that help make the context of Handles more explicit.
+pub type ParentHandle = Handle;
```

Comment:
_**Copied over comment on PR #207 so we can work on it here**_

The alias doesn’t provide any type safety. That might be okay. But if you want to have type safety (e.g., you can’t use a `ParentHandle` where a `ChildHandle` is expected), you can do this without separate types by using generics:

If you define the handle type like so:

```
pub struct Handle<Cat> {
    name: Arc<str>,
    marker: std::marker::PhantomData<Cat>,
}
```

The `Cat` type argument isn’t actually used, it really is only a marker to be used with marker structs, like so:

```
pub struct Parent;
pub type ParentHandle = Handle<Parent>;

pub struct Child;
pub type ChildHandle = Handle<Child>;
```

Now `ParentHandle` and `ChildHandle` are distinct types but you only need to define all the methods and whatnot once.

Reply:
Pushed! Took a while to update krill - but it's all working now.